### PR TITLE
fix: compute range from 1st non-whitespace word, if start==end

### LINKED ISSUE #125: Improve diagnostic range conversion when range.start=range.end
Given a go.mod file with an invalid statement, using gopls, error rendering looks bad:

Screenshot 2024-01-30 at 09 30 03

compared to vscode, where the entire words are underlined:
Screenshot 2024-01-30 at 09 31 09

### PR DESCRIPTION
**Pull request by fbricon**
Something fishy with intermediate diagnostics being computed on stale lines
Feb-20-2024 12-16-06

Fixes #125

### CODE REVIEW

File: src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java [hunk 1]
```diff
@@ -372,9 +373,31 @@ public static Range toRange(TextRange range, Document document) {
...
+    }
+
+    private static int getNextNonWhitespaceOffset(@NotNull Document document, int line, int offset) {
+        // since this method is private, at this point we already guarantee that:
+        // - line is valid
+        // - offset < document.getTextLength()
+        int textLength = document.getTextLength();
+
+        for (; offset <= document.getLineEndOffset(line) && offset < textLength; offset++) {
+            if (Character.isWhitespace(document.getText().charAt(offset))) {
```
Comment (angelozerr) [inline src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java:392]: Don't use `document.getText()` which create a new string but use **document.getCharsSequence().charAt(i);**

Comment (fbricon) [reply src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java]: nice catch!

**Comment by angelozerr** [comment]
I have written a similar code at https://github.com/redhat-developer/lsp4ij/blob/265a2d2b438642a1de2911e52c8e68073e81de6c/src/main/java/com/redhat/devtools/lsp4ij/usages/LSPUsageTargetProvider.java#L49 

So I wonder if we could have a new static method utils which retrieve a TextRange of a token just by giving an offset.

For instance `LSP4IJUtils#tokenRange(int offset, Document document)`

**Comment by angelozerr** [comment]
> Something fishy with intermediate diagnostics being computed on stale lines

This problem comes from with Go language server which is very slow to publish diagnostic. 

I don't know how to fix this problem in IJ.

**Comment by fbricon** [comment]
> So I wonder if we could have a new static method utils which retrieve a TextRange of a token just by giving an offset.
> 
> For instance `LSP4IJUtils#tokenRange(int offset, Document document)`

Done

Hopefully I didn't break anything

Screenshot 2024-03-14 at 17 21 34

File: src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java [hunk 2]
```diff
@@ -345,9 +349,43 @@ public static Range toRange(TextRange range, Document document) {
...
 
-    private static boolean isEndOfLine(int offset, int line, @NotNull Document document) {
-        int lineEndOffset = document.getLineEndOffset(line);
-        return offset == lineEndOffset;
+    private static boolean isEndOfLine(@NotNull Document document, int line, int offset) {
+        return offset == document.getLineEndOffset(line);
+    }
+    
+    @Nullable
+    public static TextRange getTokenRange(Document document , int offset) {
```
Comment (angelozerr) [inline src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java:357]: Add javadoc

**Comment by angelozerr** [comment]
Do you think you could add some test?

**Comment by angelozerr** [comment]
It works like a charm,thanks @fbricon !

**Pull request merged**